osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace osu.Game.Rulesets.Osu.Objects.Drawables
+{
+    /// <summary>
+    /// Stores the spinning history of a single spinner.<br />
+    /// Instants of movement deltas may be added or removed from this in order to calculate the total rotation for the spinner.
+    /// </summary>
+    /// <remarks>
+    /// A single, full rotation of the spinner is defined as a 360-degree rotation of the spinner, starting from 0, going in a single direction.<br />
+    /// </remarks>
+    /// <example>
+    /// If the player spins 90-degrees clockwise, then changes direction, they need to spin 90-degrees counter-clockwise to return to 0
+    /// and then continue rotating the spinner for another 360-degrees in the same direction.
+    /// </example>
+    public class SpinnerSpinHistory
+    {
+        /// <summary>
+        /// The sum of all complete spins and any current partial spin, in degrees.
+        /// </summary>
+        /// <remarks>
+        /// This is the final scoring value.
+        /// </remarks>
+        public float TotalRotation => 360 * completedSpins.Count + currentSpinMaxRotation;
+
+        private readonly Stack<CompletedSpin> completedSpins = new Stack<CompletedSpin>();
+
+        /// <summary>
+        /// The total accumulated rotation.
+        /// </summary>
+        private float totalAbsoluteRotation;
+
+        private float totalAbsoluteRotationAtLastCompletion;
```

Comment:
So this isn't "absolute" rotation, it's just accumulated rotation? Maybe rename `totalAccumulatedRotation`/`totalAccumulatedRotationAtLastCompletion`?

Reply:
Yep, can go with that. I was having a hard time naming this because it's both accumulated and absolute.